### PR TITLE
Move ImageService.cs to Jellyfin.Api

### PR DESCRIPTION
**Pull request by crobibero**

### CODE REVIEW

**Comment by Ullmie02** [comment]
I came up with a possible solution which may be a good starting point:
If you add the following instead of the TODO:

            var imagePath = imageResult.Item1;
            var imageContentType = imageResult.Item2;
            var dateImageModified = imageResult.Item3;

            bool disableCaching = Request.Headers[HeaderNames.CacheControl].Contains("no-cache");
            bool parsingSuccessful = DateTime.TryParseExact(Request.Headers[HeaderNames.IfModifiedSince], "ddd, dd MMM yyyy HH:mm:ss \"GMT\"", new CultureInfo("en-US", false), DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal, out DateTime ifModifiedSinceHeader);

            // if the parsing of the IfModifiedSince header was not successfull, disable caching
            if (!parsingSuccessful)
            {
                // disableCaching = true;
            }

            foreach (KeyValuePair<string, string> header in headers)
            {
                Response.Headers.Add(header.Key, header.Value);
            }

            Response.ContentType = imageContentType;
            Response.Headers.Add(HeaderNames.Age, Convert.ToInt64((DateTime.UtcNow - imageResult.Item3).TotalSeconds).ToString(CultureInfo.InvariantCulture));
            Response.Headers.Add(HeaderNames.Vary, HeaderNames.Accept);

            if (disableCaching)
            {
                Response.Headers.Add(HeaderNames.CacheControl, "no-cache, no-store, must-revalidate");
                Response.Headers.Add(HeaderNames.Pragma, "no-cache, no-store, must-revalidate");
            }
            else
            {
                if (cacheDuration.HasValue)
                {
                    Response.Headers.Add(HeaderNames.CacheControl, "public, max-age=" + cacheDuration.Value.TotalSeconds);
                }
                else
                {
                    Response.Headers.Add(HeaderNames.CacheControl, "public");
                }

                Response.Headers.Add(HeaderNames.LastModified, dateImageModified.ToUniversalTime().ToString("ddd, dd MMM yyyy HH:mm:ss \"GMT\"", new CultureInfo("en-US", false)));

                // if the image was not modified since "ifModifiedSinceHeader"-header, return a HTTP status code 304 not modified
                if (!(dateImageModified > ifModifiedSinceHeader))
                {
                    if (ifModifiedSinceHeader.Add(cacheDuration!.Value) < DateTime.UtcNow)
                    {
                        Response.StatusCode = StatusCodes.Status304NotModified;
                        return new ContentResult();
                    }
                }
            }

            // if the request is a head request, return a NoContent result with the same headers as it would with a GET request
            if (isHeadRequest)
            {
                return NoContent();
            }

            var stream = new FileStream(imagePath, FileMode.Open, FileAccess.Read);
            return File(stream, imageContentType);`

the web client works but caching clearly doesn't. If you scroll in the library, the images get downloaded over and over again. I haven't done much with caching, so I haven't expected anything else.
I think the problem is, that the IfModifiedSince header isn't parsed correctly so I have commented out the line `disableCaching = true;` but it didn't help.


In order to test the endpoint, `backgroundColor`, `foregroundColor`, `format `have to be nullable and you have to either fix the warnings or disable `TreatWarningsAsErrors `in `Jellyfin.Api.csproj`.

**Comment by crobibero** [comment]
I'm not worries about caching not seeming to work properly, it doesn't look like web even utilizes those headers.
I tested browsing around and always had an image, so I think this PR is good to go!

File: Jellyfin.Api/Controllers/ImageController.cs [hunk 1]
```diff
@@ -0,0 +1,1295 @@
...
+        }
+
+        /// <summary>
+        /// Set item image.
+        /// </summary>
+        /// <param name="itemId">Item id.</param>
+        /// <param name="imageType">Image type.</param>
+        /// <param name="imageIndex">(Unused) Image index.</param>
+        /// <response code="204">Image saved.</response>
+        /// <response code="400">Item not found.</response>
```
Comment (Ullmie02) [inline Jellyfin.Api/Controllers/ImageController.cs:191]:
```suggestion
        /// <response code="404">Item not found.</response>
```

File: Jellyfin.Api/Controllers/ImageController.cs [hunk 2]
```diff
@@ -0,0 +1,1295 @@
...
+        /// <param name="userId">User Id.</param>
+        /// <param name="imageType">(Unused) Image type.</param>
+        /// <param name="index">(Unused) Image index.</param>
+        /// <response code="204">Image deleted.</response>
+        /// <returns>A <see cref="NoContentResult"/>.</returns>
+        [HttpDelete("/Users/{userId}/Images/{itemType}")]
+        [HttpDelete("/Users/{userId}/Images/{itemType}/{index?}")]
+        [SuppressMessage("Microsoft.Performance", "CA1801:ReviewUnusedParameters", MessageId = "imageType", Justification = "Imported from ServiceStack")]
+        [SuppressMessage("Microsoft.Performance", "CA1801:ReviewUnusedParameters", MessageId = "index", Justification = "Imported from ServiceStack")]
+        [ProducesResponseType(StatusCodes.Status204NoContent)]
```
Comment (Ullmie02) [inline Jellyfin.Api/Controllers/ImageController.cs:130]:
```suggestion
        [ProducesResponseType(StatusCodes.Status204NoContent)]
        [ProducesResponseType(StatusCodes.Status403Forbidden)]
```

File: Jellyfin.Api/Controllers/ImageController.cs [hunk 3]
```diff
@@ -0,0 +1,1295 @@
...
+        /// <param name="userId">User Id.</param>
+        /// <param name="imageType">(Unused) Image type.</param>
+        /// <param name="index">(Unused) Image index.</param>
+        /// <response code="204">Image updated.</response>
+        /// <returns>A <see cref="NoContentResult"/>.</returns>
+        [HttpPost("/Users/{userId}/Images/{imageType}")]
+        [HttpPost("/Users/{userId}/Images/{imageType}/{index?}")]
+        [SuppressMessage("Microsoft.Performance", "CA1801:ReviewUnusedParameters", MessageId = "imageType", Justification = "Imported from ServiceStack")]
+        [SuppressMessage("Microsoft.Performance", "CA1801:ReviewUnusedParameters", MessageId = "index", Justification = "Imported from ServiceStack")]
+        public async Task<ActionResult> PostUserImage(
```
Comment (Ullmie02) [inline Jellyfin.Api/Controllers/ImageController.cs:87]: `ProducesResponseType `missing

**Pull request merged**